src/hotspot/share/services/memoryPool.cpp

```diff
@@ -113,33 +118,35 @@ instanceOop MemoryPool::get_memory_pool_instance(TRAPS) {
                            &args,
                            CHECK_NULL);
 
-    instanceOop p = (instanceOop) result.get_oop();
-    instanceHandle pool(THREAD, p);
-
-    {
-      // Get lock since another thread may have create the instance
-      MutexLocker ml(THREAD, Management_lock);
-
-      // Check if another thread has created the pool.  We reload
-      // _memory_pool_obj here because some other thread may have
-      // initialized it while we were executing the code before the lock.
-      pool_obj = Atomic::load(&_memory_pool_obj).resolve();
-      if (pool_obj != nullptr) {
-         return (instanceOop)pool_obj;
-      }
-
-      // Get the address of the object we created via call_special.
-      pool_obj = pool();
-
-      // Use store barrier to make sure the memory accesses associated
-      // with creating the pool are visible before publishing its address.
-      // The unlock will publish the store to _memory_pool_obj because
-      // it does a release first.
-      Atomic::release_store(&_memory_pool_obj, OopHandle(Universe::vm_global(), pool_obj));
+    // Verify we didn't get a null pool.  If that could happen then we'd
+    // need to return immediately rather than continuing on and recording the
+    // pool has been created.
+    oop p = result.get_oop();
+    guarantee(p != nullptr, "Pool creation returns null");
+    instanceHandle pool(THREAD, (instanceOop)p);
+
+    // Allocate global handle outside lock, to avoid any lock nesting issues
+    // with the Management_lock.
+    OopHandle pool_handle(Universe::vm_global(), pool());
+
+    // Get lock since another thread may have created and installed the instance.
+    MutexLocker ml(THREAD, Management_lock);
+
+    if (Atomic::load(&_memory_pool_obj_initialized)) {
+      // Some other thread won the race.  Release the handle we allocated and
+      // use the other one.  Relaxed load is sufficient because flag update is
+      // under the lock.
+      pool_handle.release(Universe::vm_global());
+    } else {
+      // Record the object we created via call_special.
+      _memory_pool_obj = pool_handle;
```

Comment:
Could assert that `_memory_pool_obj.is_empty()`.

Reply:
Done.